src/main/java/bio/terra/service/dataset/flight/ingest/IngestValidateRefsStep.java

```diff
@@ -39,4 +40,39 @@ public StepResult undoStep(FlightContext context) {
     // The update will update row ids that are null, so it can be restarted on failure.
     return StepResult.getStepResultSuccess();
   }
+
+  public static class InvalidRefId {
+    public String refId;
+    public String columnName;
+
+    public InvalidRefId(String refId, String columnName) {
+      this.refId = refId;
+      this.columnName = columnName;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+      InvalidRefId that = (InvalidRefId) o;
+      return Objects.equals(refId, that.refId) && Objects.equals(columnName, that.columnName);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(refId, columnName);
+    }
+
+    @Override
+    public String toString() {
+      return "InvalidRefId{"
+          + "refId='"
+          + refId
+          + '\''
```

Comment:
It's kinda weird you have a random char here.  Might be more readable as a string format.